ratelimits/transaction.go

```diff
@@ -149,26 +160,161 @@ type TransactionBuilder struct {
+}
+
 // NewTransactionBuilderFromFiles returns a new *TransactionBuilder. The
 // provided defaults and overrides paths are expected to be paths to YAML files
 // that contain the default and override limits, respectively. Overrides is
 // optional, defaults is required.
-func NewTransactionBuilderFromFiles(defaults, overrides string) (*TransactionBuilder, error) {
-	registry, err := newLimitRegistryFromFiles(defaults, overrides)
+func NewTransactionBuilderFromFiles(defaults string, overrides string, stats prometheus.Registerer, logger blog.Logger) (*TransactionBuilder, error) {
+	defaultsData, err := loadDefaultsFromFile(defaults)
 	if err != nil {
 		return nil, err
 	}
-	return &TransactionBuilder{registry}, nil
+
+	if overrides == "" {
+		return NewTransactionBuilder(defaultsData, nil, stats, logger)
+	}
+
+	refresher := func(ctx context.Context, _ prometheus.Gauge, _ blog.Logger) (Limits, error) {
+		overridesData, err := loadOverridesFromFile(overrides)
+		if err != nil {
+			return nil, err
+		}
+		return parseOverrideLimits(overridesData)
+	}
+
+	return NewTransactionBuilder(defaultsData, refresher, stats, logger)
 }
 
-// NewTransactionBuilder returns a new *TransactionBuilder. The provided
-// defaults map is expected to contain default limit data. Overrides are not
-// supported. Defaults is required.
-func NewTransactionBuilder(defaults LimitConfigs) (*TransactionBuilder, error) {
-	registry, err := newLimitRegistry(defaults, nil)
+// NewTransactionBuilder returns a new *TransactionBuilder. A defaults map is
+// required.
+func NewTransactionBuilder(defaults LimitConfigs, overrides OverridesRefresher, stats prometheus.Registerer, logger blog.Logger) (*TransactionBuilder, error) {
+	regDefaults, err := parseDefaultLimits(defaults)
 	if err != nil {
 		return nil, err
 	}
+
+	if overrides == nil {
+		overrides = func(context.Context, prometheus.Gauge, blog.Logger) (Limits, error) {
+			return nil, nil
+		}
+	}
+
+	overridesTimestamp := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "ratelimits",
+		Subsystem: "overrides",
+		Name:      "timestamp_seconds",
+		Help:      "A gauge with the last timestamp when overrides were successfully loaded",
+	})
+	stats.MustRegister(overridesTimestamp)
+
+	overridesErrors := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "ratelimits",
+		Subsystem: "overrides",
+		Name:      "errors",
+		Help:      "A gauge with the number of errors while last trying to load overrides",
+	})
+	stats.MustRegister(overridesErrors)
+
+	overridesPerLimit := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "ratelimits",
+			Subsystem: "overrides",
+			Name:      "active",
+			Help:      "A gauge with the number of overrides, partitioned by rate limit",
+		},
+		[]string{"limit"},
+	)
+	stats.MustRegister(overridesPerLimit)
+
+	registry := &limitRegistry{
+		defaults:         regDefaults,
+		refreshOverrides: overrides,
+		stats:            stats,
```

Comment:
It's unnecessary to store the `stats prometheus.Registerer` in the `limitRegistry` object because we've already registered everything we need to.